Updated to the most recent Alonzo specification

TimSheard opened the pull request:
Completed Figures 1, 5, 6, 7, 8 and 12. Ormolise.

This PR implements all the functions in the most up todate specification in the above Figures.

TimSheard committed:
Updated to the most recent Alonzo specification

Completed Figures 1, 2, 3, 4, 5, 6, 7, 8 and 12.

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:
```diff
@@ -24,30 +27,30 @@ module Cardano.Ledger.Alonzo.PParams
-    encodeMapLen,
-    encodeWord,
-  )
+import Cardano.Binary (FromCBOR (..), ToCBOR (..), decodeWord, encodeListLen, encodeMapLen, encodeWord, serialize')
```
maybe it's just me, but when the imports get past three or so, I prefer them vertical. it makes it easier to see what is changing in diffs.

TimSheard replied:
done

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:
```diff
@@ -413,3 +420,39 @@ newtype PPHash crypto
+hash2P :: forall t e. Era e => Proxy t -> Proxy e -> [HashItem] -> Hash.Hash (CC.HASH (Crypto e)) t
+hash2P _ _ x = Hash.castHash (Hash.hashWith serializeHashTuple x)
+
+class Era e => Hashable e t where
```
we are using the `Hashable` typeclass from `Data.Hashable` (hashable package) elsewhere in our codebase, so we will need a new name.

TimSheard replied:
I changed it to HashTuple, since all instances are tuples.

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -17,37 +22,80 @@ module Cardano.Ledger.Alonzo.Tx
     Tx (Tx, body, wits, isValidating, auxiliaryData),
     ScriptPurpose (..),
     Indexable (..),
+    --  Figure 5
```
this is awesome lining up these exports with the figures from the spec!

TimSheard replied:
I have done an even better job, adding new figures.

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -204,23 +252,92 @@ deriving via
+txscriptfee (Prices pr_mem pr_steps) (ExUnits mem steps) =
+  (mem <×> pr_mem) <+> (steps <×> pr_steps)
+
+-- TODO Check that txSize is the right functio-- Length of the serialised bytes
```
yep, what you defined below is correct. note that in shelley we [didn't camel case](https://github.com/input-output-hk/cardano-ledger-specs/blob/d104b97739bf37b695251b3ece5430f867fe0411/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs#L772) it (and we should be consistent).

TimSheard replied:
fixed that.

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -204,23 +252,92 @@ deriving via
+    b protparam = Coin (fromIntegral (_minfeeB protparam))
+
+-- IT IS NOT AT ALL CLEAR what "validatorHash a" means in
+-- the specification. Maybe something like this?
```
yes, this is correct. let's remove this comment before merging. we could add some prose to the spec if you think that that would help.

TimSheard replied:
done

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+              Nothing -> []
+              Just d -> [d]
+
+-- IS IT REALLY THE CASE ONLY SPENDING PURPOSE HAS DATA? TODO
```
maybe @polinavino  or @WhatisRT can confirm.

WhatisRT replied:
yes, all the other purposes are essentially state-less, so they don't have data

TimSheard replied:
fixed the comment

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+  (Certifying $ DCertDeleg c, hk) : ans
+add_only_cwitness !ans _ = ans
+
+-- TODO   Why is 'h' in the existential? h should be underscore or _h
```
we can remove this comment now.

TimSheard replied:
done

JaredCorduan commented on alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs:
```diff
@@ -98,6 +94,7 @@ newtype IsFee = IsFee Bool
       FromCBOR
     )
 
+{-
```
since `TxIn` will be the same in Shelley and Alonzo, can't we delete this commented code?

nc6 replied:
Even if it won't, git keeps history! So we can always delete commented code

TimSheard replied:
done

JaredCorduan reviewed (approved):
This is a bunch of the alonzo figures, all lining up so nicely with the spec, it's looking really great! I made a few comments about mostly trivial things.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:
```diff
@@ -75,6 +78,7 @@ import Shelley.Spec.Ledger.Serialization
   )
 import Shelley.Spec.Ledger.Slot (EpochNo (..))
 
+-- ================================================================
```
Very minor, but I find these things kind of annoying, and I know for example that they end up being removed in other PRs. Do you find them that helpful?

TimSheard replied:
Yes, one should always separate the imports from the code. Our files sometimes have a hundred or more lines of imports.  This allows you to find where the code starts with a glance.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:
```diff
@@ -401,6 +405,9 @@ updatePParams pp ppup =
     fromMaybe' x = fromMaybe x . strictMaybeToMaybe
 
+-- ===================================================
+-- Figure 1: "Definitions Used in Protocol Parameters"
```
I think if you're going to link to the doc, we need to make the link more explicit

TimSheard replied:
OK I added this documentation to the module
This module exports implementations of many of the functions outlined in the Alonzo specification.
     The link to source of the specification
       https://github.com/input-output-hk/cardano-ledger-specs/tree/master/alonzo/formal-spec
     The most recent version of the document can be found here:
       https://hydra.iohk.io/job/Cardano/cardano-ledger-specs/specs.alonzo-ledger/latest/download-by-type/doc-pdf/alonzo-changes
     The functions can be found in Figures in that document, and sections of this code refer to those figures.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:
```diff
@@ -413,3 +420,39 @@ newtype PPHash crypto
 deriving newtype instance CC.Crypto crypto => ToCBOR (PPHash crypto)
+
+hashLanguagePP :: forall era. Era era => PParams era -> Language -> PPHash (Crypto era)
+hashLanguagePP pp lang = PPHash (hash2P (Proxy @EraIndependentPP) (Proxy @era) [HashItem restrictedmap])
```
Very long line, can we try to keep lines to 80 chars?

TimSheard replied:
I fixed this.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:
```diff
@@ -413,3 +420,39 @@ newtype PPHash crypto
+    restrictedmap = eval (singleton lang ◁ (_costmdls pp))
+
+-- ===============================================================
+-- Hashing Tuples (upto to width 4) in a uniform way
```
I'm not convinced that this is something we want? We generally err on the side of _not_ doing things in such a uniform way (e.g. not using generic instances for serialisation) because we want control over how exactly we do so in areas which will affect hashes/validation of data stored on the chain.

We also don't use this functionality outside of this module, so it's not clear what the advantage of it is - it adds a new abstraction that somewhat duplicates existing abstractions but which isn't widely used.

TimSheard replied:
I think we need some discussion about these things. The Alonzo spec is hashing al sorts of stuff. Ideally we would have some practical way to ensure that users can construct equivalent hashes. This is going to need some thought.

nc6 replied:
Happy to have a discussion about this and work out a sensible abstraction, but let's not introduce it in an unrelated module in an unrelated PR! Please go ahead and create a ticket that explains the problem here and then we can think about how to solve it.

TimSheard replied:
I have removed this. instead we use the strategy that anything hashed is a Memobytes.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -17,37 +22,80 @@ module Cardano.Ledger.Alonzo.Tx
     ScriptPurpose (..),
     Indexable (..),
+    --  Figure 5
+    txinputs_fee,
```
Why snake case? We typically reserve this for c interaction functions.

JaredCorduan replied:
my guess is that this was done to match the actual subscript in the spec, which is IMHO reasonable.

TimSheard replied:
This is an attempt to mimic the subscripts that are used in the specification.

TimSheard replied:
Renamed things consistent with other Era

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -204,23 +252,92 @@ deriving via
+-- From the specification, Figure 5 "Functions related to fees"
+-- ===============================================================
+
+isNonNativeScriptAddress :: forall era. ValidateScript era => Tx era -> Addr (Crypto era) -> Bool
```
[style] line length

TimSheard replied:
fixed this

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -204,23 +252,92 @@ deriving via
+txSize :: Tx era -> Integer
+txSize (TxConstr (Memo _ bytes)) = fromIntegral (SBS.length bytes)
+
+txinputs_fee :: AlonzoBody era => TxBody era -> Set (TxIn (Crypto era))
```
This is just a field, we shouldn't need a new function for it

TimSheard replied:
just mimicing the specification.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -17,37 +22,80 @@ module Cardano.Ledger.Alonzo.Tx
 import Data.MemoBytes (Mem, MemoBytes (Memo), memoBytes)
 import Data.Sequence.Strict (StrictSeq)
 import qualified Data.Sequence.Strict as StrictSeq
-import qualified Data.Set as Set
```
Data.Set recommends (https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Set.html) qualified import, and this is consistent with the rest of the codebase. So please don't make it different here!

TimSheard replied:
done

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+-- ===============================================================
+
+-- | valContext collects info from the Tx and the UTxO and translates it into
+--   A Data, which the Plutus language know how to interpret.
```
```suggestion
--   a 'Data', which the Plutus language know how to interpret.
```

TimSheard replied:
done

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+valContext :: UTxO era -> Tx era -> ScriptPurpose (Crypto era) -> [Data era]
+valContext _utxo _tx _sp = []
+
+--TODO FIX THIS, when defined will always return singleton list
```
?

TimSheard replied:
This function has to wait until we have the Plutus API. Just a note to remind us to fix it when we have that.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+    (IsValidating b, _exunits) = runPLCScript cost AlonzoScript.PlutusScript ds units
+
+-- ===================================================================
+-- From Specification, Figure 12 "UTXOW helper functions"
```
I feel like we might want to eventually move these elsewhere, to avoid `Tx` becoming a giant bag of all functions (a la `LedgerState.hs`), but we can do that subsequently

TimSheard replied:
Certainly possibility

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+
+-- We only find certificate witnesses in Delegating and Deregistration DCerts
+-- that have ScriptHashObj credentials.
+add_only_cwitness ::
```
snake_case?

TimSheard replied:
fixed

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+txscripts x = witsScript x
+
+txwits :: (Era era, ToCBOR (Core.AuxiliaryData era)) => Tx era -> TxWitness era
+txwits x = wits x
```
Why this duplicate function?

TimSheard replied:
mimicing the specification.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
+  (Era era, ToCBOR (Core.Script era)) =>
+  TxWitness era ->
+  Map.Map (ScriptHash (Crypto era)) (Core.Script era)
+txscripts x = witsScript x
```
Duplicate?

TimSheard replied:
Trying to keep the code consistent with the spec.  The other thing to do would be to change the names of the Haskell accessor functions.

nc6 replied:
Gotcha. I think in this case it might be best to change the spec, to be honest!

TimSheard replied:
I will make some changes to the spec when Andre has fixed up his changes.

nc6 commented on semantics/small-steps-test/small-steps-test.cabal:
```diff
@@ -33,6 +33,7 @@ library
                      , microlens-th
                      , mtl
                      , nothunks
+                     , prettyprinter
```
How is this related to this PR?

TimSheard replied:
It seems the tests won't build without it. See Issue 2093, and the transcript of the build here
https://github.com/input-output-hk/cardano-ledger-specs/issues/2093

nc6 reviewed (changes requested):
A bunch of minor things

TimSheard commented:
This latest push does a few things
1) It cleans up TxWitness so that the subpart that must be hashed is a public Memobytes. I also added a new pattern FlatWitness that lets users access it in the old style where there we 5 fields. So one can use either view.

2) I tried to make everything that needed to be hashed a Memobytes

3) I did some renaming of fields to be consistent and added HasField instances identedical to earlier errors.

4) There are still two places where hashing need to be further addressed. in PParams and the function hashScriptData.  I have stubbed out these parts. I suggest we get this PR merged and go back to address those issues.

nc6 reviewed (approved):
Looks great, thanks!
